src/main/java/graphql/schema/validation/TypeAndFieldRule.java

```diff
@@ -0,0 +1,274 @@
+        if (fieldName.length() >= 2 && fieldName.startsWith("__")) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.InvalidCustomizedNameError,
+                    String.format("\"%s\" in \"%s\" must not begin with \"__\", which is reserved by GraphQL introspection.", fieldName, typeName));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private void assertArgumentName(String typeName, String fieldName, String argumentName, SchemaValidationErrorCollector errorCollector) {
+        if (argumentName.length() >= 2 && argumentName.startsWith("__")) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.InvalidCustomizedNameError,
+                    String.format("Argument name \"%s\" in \"%s-%s\" must not begin with \"__\", which is reserved by GraphQL introspection.", argumentName, typeName, fieldName));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private void assertEnumValueDefinitionName(String typeName, String enumValueDefinitionName, SchemaValidationErrorCollector errorCollector) {
+        if (enumValueDefinitionName.length() >= 2 && enumValueDefinitionName.startsWith("__")) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.InvalidCustomizedNameError,
+                    String.format("EnumValueDefinition \"%s\" in  \"%s\"  must not begin with \"__\", which is reserved by GraphQL introspection.", enumValueDefinitionName, typeName));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private void assertNonNullType(GraphQLType type, SchemaValidationErrorCollector errorCollector) {
+        if (type instanceof GraphQLNonNull && ((GraphQLNonNull) type).getWrappedType() instanceof GraphQLNonNull) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.NonNullWrapNonNullError,
+                    String.format("Non‐Null type \"%s\" must not wrap another Non‐Null type.", GraphQLTypeUtil.simplePrint(type)));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private List<GraphQLNamedType> filterScalarAndIntrospectionTypes(List<GraphQLNamedType> graphQLNamedTypes) {
+        if (graphQLNamedTypes == null || graphQLNamedTypes.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        return graphQLNamedTypes.stream().filter(type -> !isScalarOrIntrospectionTypes(type)).collect(Collectors.toList());
+    }
+
+    private boolean isScalarOrIntrospectionTypes(GraphQLNamedType type) {
+        if (type instanceof GraphQLScalarType && (
+                type.getName().equals("Int") || type.getName().equals("Float") || type.getName().equals("String") ||
```

Comment:
we should only treat the build-in scalar types (Int, Float, String, Boolean and ID) in a special way. The others are not really special and should be treated as every other custom scalar.

We also have a helper method in Scalars (I believe) to determine if the type is a build-in one or not

Reply:
Thanks for your suggestion.